Create API Key on behalf of other user

tvernum opened the pull request:
This change adds a "grant API key action"

   POST /_security/api_key/grant

that creates a new API key using the privileges of one user ("the
system user") to execute the action, but creates the API key with
the roles of the second user ("the end user").

This allows a system (such as Kibana) to create API keys representing
the identity and access of an authenticated user without requiring
that user to have permission to create API keys on their own.

Relates: #48716

tvernum committed:
Create API Key on behalf of other user

This change adds a "grant API key action"

   POST /_security/api_key/grant

that creates a new API key using the privileges of one user ("the
system user") to execute the action, but creates the API key with
the roles of the second user ("the end user").

This allows a system (such as Kibana) to create API keys representing
the identity and access of an authenticated user without requiring
that user to have permission to create API keys on their own.

elasticmachine commented:
Pinging @elastic/es-security (:Security/Authentication)

tvernum commented:
This lacks:
- docs
- rest spec
- rest (yaml) tests
- HLRC
- a new privilege (`grant_api_key`)

but it's big enough as it is, and I want to keep it to a reviewable size.

tvernum committed:
Revert formatting change

tvernum committed:
Revert unnecessary changes

tvernum commented:
Ping: @peterschretlen @mikecote

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/ApiKeyGenerator.java:
```diff
@@ -0,0 +1,58 @@
+            listener.onFailure(new ElasticsearchSecurityException("no authentication available to generate API key"));
+            return;
+        }
+        apiKeyService.ensureEnabled();
```
This is called in `ApiKeyService#createApiKey()`. Is it intentional that we do it here too becayse `getRoleDescriptors` can be expensive ? If so can we also /instead inject the ApiKeyService in the Transport service and short circuit there ?

tvernum replied:
Partly the relative expense, but partly the "right" error message. If the API key service is disabled you want to prioritise that error message ahead of any role format issues.
(That's specifically what triggered me to write it, I ran into something along those lines while testing)

tvernum replied:
> If so can we also /instead inject the ApiKeyService in the Transport service and short circuit there

I don't know if we want to. I think we should prioritise authentication problems (even secondary auth problems) over the service being disabled. Happy to discuss if you feel otherwise.

jkakavas replied:
> , but partly the "right" error message.

Makes sense. 

> I think we should prioritise authentication problems 

Agreed, I didn't think this through

jkakavas reviewed (commented):
Did a first pass but didn't look that closely into tests. Will make another pass early next week

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/GrantApiKeyRequest.java:
```diff
@@ -0,0 +1,171 @@
+    @Override
+    public ActionRequestValidationException validate() {
+        ActionRequestValidationException validationException = apiKey.validate();
+        if (grant.type == null) {
```
Can do, but is there a reason why you'd like that?

jkakavas replied:
Not the end of the world as we know it, I think it reads better, feel free to decide.

tvernum commented on x-pack/plugin/security/qa/security-basic/src/test/java/org/elasticsearch/xpack/security/apikey/ApiKeyRestIT.java:
```diff
@@ -0,0 +1,80 @@
+        invalidateApiKeysForUser(END_USER);
+    }
+
+    public void testGrantApiKeyForOtherUser() throws IOException {
```
I can't in this test, because basic license doesn't suppor tokens, but I can move the test elsewhere and do that if you'd like it.

Since the transport action test already covers Username+Password and AccessToken (both valid and invalid), I didn't think it was necessary to cover it again in the IT. I just wanted something integrated to test it top-to-bottom.

Testing's a tricky balance, so if it's something you think we need, I'm happy to make it happen (but it will probably means a whole new QA project for rest tests using  security on trial licenses).

jkakavas replied:
SGTM to leave it as is

albertzaharovits replied:
> I can't in this test, because basic license doesn't support tokens

But you can start a trial, as it's done in `SecurityWithBasicLicenseIT#testWithTrialLicense` , right ? If this is possible, I would also just slightly prefer a test case using tokens.

tvernum replied:
No, because a cluster can only start a trial once, and the other test does that.
If we want to test this with a non-basic license then we'll need a new QA project.

tvernum replied:
You convinced me that I was just being lazy, so I've added a new QA project, and a new test for authenticating with a token + API key expiration.

mikecote commented:
Hi @tvernum I played with this PR locally last Friday. Thank you for getting this done 🙏 From my perspective it works great and I can see what needs to be done on Kibana side without issue. As you noted, the privileges aren't given to the kibana user so I did my testing with `elastic` / others with `manage_api_key` privilege for now and it works! 👍

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/TransportGrantApiKeyAction.java:
```diff
@@ -0,0 +1,86 @@
+                return;
+            default:
+                listener.onFailure(new ElasticsearchSecurityException("the grant type [{}] is not supported", grant.getType()));
+                return;
```
The `return` ? Not strictly, but I did it for symmetry with the other switch cases.
Does it bother you? I'm happy to change it, and I'm personally not consistent in my balance between "symmetry" and "necessity"

jkakavas replied:
again, this was a nit, feel free to address or not. I tend to review ( also not very consistently I'd admit ) taking personal preference into consideration and I usually don't hold strong opinions on such things.

tvernum committed:
Merge branch 'master' into feature/48716-api-key-other-user

tvernum committed:
Move API key test to new QA/security-trial project

This test was piggy-backing on the security-in-basic QA project we
had, but the grant-api-key endpoint has the ability to use tokens,
which are not a basic licensed feature.

This creates a new QA project for security on trial licenses and run
the API key tests there

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:
```diff
@@ -394,6 +394,32 @@ void getAndValidateToken(ThreadContext ctx, ActionListener<UserToken> listener)
+     * {@code null} authentication object.
+     */
+    public void authenticateToken(SecureString tokenString, ActionListener<Authentication> listener) {
+        if (isEnabled()) {
```
I believe this new method should return the new [fancy 400 exceptions](https://github.com/elastic/elasticsearch/pull/52811) when the token service is not enabled (since it's not used in the authentication chain).

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java:
```diff
@@ -394,6 +394,32 @@ void getAndValidateToken(ThreadContext ctx, ActionListener<UserToken> listener)
+                        listener::onFailure
+                    ));
+                } else {
+                    listener.onResponse(null);
```
Also because this new method is not used in the authn chain, when a token fails decoding it's a client error.
I believe:
```
listener.onFailure(new IllegalArgumentException());
```
would do, it would return 400 (currently this falls back to 500 in `ApiKeyGenerator#generateApiKey`).

jkakavas replied:
Good point @albertzaharovits

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/apikey/RestGrantApiKeyAction.java:
```diff
@@ -0,0 +1,128 @@
+        try (XContentParser parser = request.contentParser()) {
+            final GrantApiKeyRequest grantRequest = PARSER.parse(parser, null);
+            if (refresh != null) {
+                grantRequest.setRefreshPolicy(WriteRequest.RefreshPolicy.parse(refresh));
```
The `refresh_policy` should be forwarded to the inner `CreateApiKeyRequest` to take effect.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/apikey/RestGrantApiKeyAction.java:
```diff
@@ -0,0 +1,128 @@
+        public void onFailure(Exception e) {
+            RestStatus status = ExceptionsHelper.status(e);
+            if (status == RestStatus.UNAUTHORIZED) {
+                sendFailure(new ElasticsearchSecurityException("Failed to authenticate api key grant", RestStatus.FORBIDDEN, e));
```
I like this wrapping to translate 401 to 403!
Assuming this is the sole reason of existence for the inner class, in order to make it easier to read, I would make it more concise:
```
            return channel -> client.execute(GrantApiKeyAction.INSTANCE, grantRequest,
                    ActionListener.delegateResponse(new RestToXContentListener<>(channel),
                            (l, e) -> {
                                RestStatus status = ExceptionsHelper.status(e);
                                if (status == RestStatus.UNAUTHORIZED) {
                                    l.onFailure(new ElasticsearchSecurityException("Failed to authenticate api key grant",
                                            RestStatus.FORBIDDEN, e));
                                } else {
                                    l.onFailure(e);
                                }
                            }));
```

tvernum replied:
Thanks @albertzaharovits that was a huge improvement.

albertzaharovits reviewed (approved):
LGTM overall, I spotted only minor problems, no second take is necessary after we settle those.

jkakavas reviewed (approved):
LGTM Tim

tvernum committed:
Merge branch 'master' into feature/48716-api-key-other-user

tvernum committed:
Address feedback

tvernum committed:
Merge branch 'master' into feature/48716-api-key-other-user

tvernum committed:
Address more feedback
